accton.8
.TH ACCTON 8 "2008 November 24"
.SH NAME
accton \-  turns process accounting on or off
.SH SYNOPSIS
.hy 0
.na
.TP
.B accton
.RI [\| OPTION \|]
.RB on \||\| off \||\| filename
.TP
.B accton
[
.B \-V
|
.B \-\-version 
]
[
.B \-h
|
.B \-\-help
]
.SH DESCRIPTION
.LP
.B accton
.I filename
turns on process accounting.  If called with no arguments,
it will, by default, stop process accounting.
.SH OPTIONS
.PD 0
.TP
.TP
.B \-V, \-\-version
Print the version number of 
.B ac
to standard output and quit.
.TP
.B \-h, \-\-help
Prints the usage string and default locations of system files to
standard output and exits.
.SH FILES
.TP
.I acct
The system wide process accounting file. See
.BR acct (5)
(or
.BR pacct (5))
for further details.
.LP
.SH AUTHOR
The GNU accounting utilities were written by Noel Cragg
<noel@gnu.ai.mit.edu>. The man page was adapted from the accounting
texinfo page by Susan Kleinmann <sgk@sgk.tiac.net>.
.SH "SEE ALSO"
.BR acct (5),
.BR ac (1)


NEWS
New in 6.5.3:

  * Merged patches from Debian and Red Hat.

New in 6.5.1:

  * GNUlib for compability, audited code... Reworked.

New in 6.5:

  * Patches from Debian.

  * Actually released to the public.

  * Huge code-refactoring.

New in 6.4:

  * Linux multiformat support: Starting from release 2.6.8, the Linux
    kernel writes either the totally new v3 format or v1 or v2 formats
    that are enhanced relative to the v0 format of earlier kernels.
    GNU acct now understands v0, v2, and v3 formats (v1, written on
    the m68k architecture, is not supported) and uses the information
    from the additional fields in v2 format. Particularely, 32 bit
    user and group IDs are supported.

  * Added `--format', `--byteswap', and `--ahz' options to dump-acct.
    Together with the existing `--raw' switch and Linux multiformat
    support this results in a handy format converter.

  * The acct.h header file is now expected to be at <sys/acct.h>,
    the older <linux/acct.h> is ignored. As this file is not used
    at all when Linux multiformat is enabled (which is the default),
    noone should notice this change anyways.

  * A bug with the ctime() function on 64 bit platforms was fixed.

  * Some small code cleanups were performed, GNU acct should now compile
    without warnings on ANSI C compilers.

  * The build system was updated to recent versions of the tools used.


New in 6.3.5:
  * Reverted below change to acct.h header search.


New in 6.3.4:
  * When trying to find the acct.h header file, look in the linux
    directory before the sys directory, since the former is probably
    newer than the latter.

  * Documentation fixes for sa.


New in 6.3.3:

  * For those systems which have wide utmp fields, last now tries to
    fit its output into 80 columns by using "reasonable" default
    fields lengths for those members.  The `-w' and `--wide' flags
    have been added to allow you to view the fields with their "full
    precision," as it were.

  * Fix elapsed time calculation in sa.

  * Added an `--ahz' flag to lastcomm and sa to make it easier for
    those folks who are trying to analyze logs generated on other
    systems with the same byte order and file format but a different
    value for AHZ.

  * Fixed the `--strict-match' flag in lastcomm.  The following now
    works as expected:

      lastcomm --strict-match noel cat ttyp7


New in 6.3.2:

  * Small memory allocation problem fixed in sa.

  * Released to the general public.


New in 6.3.1:

  * A `NEWS' file (and others) so that we conform to the GNU
    requirements for packaging.

  * We use `automake' to build the package, so things like ansi2knr
    should be called automatically for those of you that don't have
    ANSI C compilers.  It also helps write the install targets, etc.
    Very cool.  I was suprised how easy it was to integrate into this
    package.  Hats off to David J. MacKenzie and Tom Tromey!


New in 6.3:

  * Hey!  We're on prep.ai.mit.edu!

  * Renamed long option `--year' to `--print-year' in both the ac and
  last programs.

  * Added a `--print-seconds' (or `-s') flag to last, for those who
  really want the nitty-gritty.

  * Added a `--sort-real-time' option to sa for those who want to sort
  the output by "real" (elapsed) time.

  * Fixed a buffer overrun error in sa which was preventing it from
  sorting the username list correctly.

Ancient history:

  * Check the file `ChangeLog' in the source directory.


accton.c
/*
Copyright (C) 1993, 1996, 1997, 2003, 2008, 2009, 2010 Free Software Foundation, Inc.

This file contains the code for GNU accton.

The GNU Accounting Utilities are free software; you can redistribute
them and/or modify them under the terms of the GNU General Public
License as published by the Free Software Foundation; either version
3, or (at your option) any later version.

The GNU Accounting Utilities are distributed in the hope that they will
be useful, but WITHOUT ANY WARRANTY; without even the implied warranty
of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with the GNU Accounting Utilities; see the file COPYING.  If
not, write to the Free Software Foundation, 675 Mass Ave, Cambridge,
MA 02139, USA.
*/

#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#ifdef HAVE_ERRNO_H
#include <errno.h>
#else
extern int errno;
#endif
#include <string.h>

#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif

#include "common.h"
#include "files.h"
#ifdef HAVE_GETOPT_LONG_ONLY
#include <getopt.h>
#else
#include "getopt.h"
#endif
#include "version.h"

/* globals */

char *program_name;		/* name of this executable */

void print_option_error (void);
void print_version (void);
void print_help (void);

/* code */

/*
 * to print when there are errors related to options
 */

void print_option_error(void)
{
  (void)printf ("Try '%s --help' for more information.\n", program_name);
}

/*
 * to print version
 */

void print_version(void)
{
  (void)printf("%s: GNU Accounting Utilities (release %s)\n",program_name, VERSION_STRING);
}

/*
 * to print usage (kind of very brief help)
 */

void give_usage(void)
{
  (void)printf ("Usage: %s [OPTION] on|off|ACCOUNTING_FILE\n", program_name);
}

/*
 * to print help
 */

void print_help(void)
{
  give_usage ();

  printf ("\n\
          Turns process accounting on or off, or changes the file where this\n\
          info is saved.\n\n\
          OPTIONS:\n\
          -h, --help       Show help and exit\n\
          -V, --version    Show version and exit\n\n\
          ARGUMENTS:\n\
          on               Activate process accounting and use default file\n\
          off              Deactivate process accounting\n\
          ACCOUNTING_FILE  Activate (if not active) and save information in\n\
          this file\n\n\
          The system's default process accounting file is '%s'.\n\n\
          Report bugs to <bug-acct@gnu.org>\n",
          ACCT_FILE_LOC);
}


/*
 * does all the real work
 */

int main(int argc, char *argv[])
{
  int c, status_acct, number_args;
  char *acct_file = NULL;

  program_name = argv[0];


  /* deal with options */
  while (1)
    {
      int option_index = 0;

      static struct option long_options[] =
      {
        { "version", no_argument, NULL, 1
        },
        { "help", no_argument, NULL, 2 },
        { 0, 0, 0, 0 }
      };

      c = getopt_long (argc, argv, "hV", long_options, &option_index);

      if (c == EOF)
        break;

      switch (c)
        {
        case 'V':
        case 1:
          print_version ();
          exit (EXIT_SUCCESS);
        case 'h':
        case 2:
          print_help ();
          exit (EXIT_SUCCESS);
        default:
          give_usage ();
          print_option_error ();
          exit (EXIT_FAILURE);
        }
    }


  /* get the number of real arguments -- not options nor prog. name */
  number_args = argc - optind;

  /* more than one arg -> error, too many args */
  if (1 < number_args)
    {
      printf ("%s: too many arguments\n", program_name);
      give_usage ();
      print_option_error ();
      exit (EXIT_FAILURE);
    }
  /* no arguments -> error, no arguments */
  else if (0 == number_args)
    {
      printf ("%s: no arguments\n", program_name);
      give_usage ();
      print_option_error ();
      exit (EXIT_FAILURE);
    }
  /* exactly one arg -> on|off|ACCOUNTING_FILE */
  else
    {
      acct_file = argv[optind];

      /* if the argument was "on" then pass the default file to activate.. */
      if (0 == strcasecmp (acct_file, "on"))
        {
          status_acct  = acct (ACCT_FILE_LOC);
          printf ("Turning on process accounting, file set to the default '%s'.\n",
                  ACCT_FILE_LOC);
        }
      /* ..else if was "off", pass NULL to deactivate.. */
      else if (0 == strcasecmp (acct_file, "off"))
        {
          status_acct  = acct (NULL);
          printf ("Turning off process accounting.\n");
        }
      /* ..else, the argument will be used as filename */
      else
        {
          status_acct  = acct (acct_file);
          printf ("Turning on process accounting, file set to '%s'.\n", acct_file);
        }


      /* did the system call return an error? */
      if (-1 == status_acct)
        {
          int err = errno;
          perror ("accton");
          exit (err);
        }
    }
  return EXIT_SUCCESS;
}

